### DISCUSSION TITLE
apachectl graceful crashes httpd on FreeBSD with mod_auth_openidc loaded

### DISCUSSION THREAD

**Discussion by pk1048** [OpenIDC/mod_auth_openidc#980] [Q&A]
We are running FreeBSD 12.3 and apache 2.4.54 with mod_auth_openidc 2.4.7 (I know, an old release). An apachectl restart works fine but a graceful causes the httpd service to crash with the following error.
`[Fri Dec 16 18:46:12.095799 2022] [auth_openidc:error] [pid 21372] oidc_cache_mutex_post_config: apr_global_mutex_create failed to create mutex on file /tmp/mod_auth_openidc_shm_mutex.21372.800b2e110: File exists (17)`

This looks like the shared memory mutex is not being completely closed when the thread is signalled to exit (as part of the graceful).

Note that we see this whether using shm (default) or memcache for the cache.

I did not see any other discussion of graceful and mutex issues since the fix in 2.4.6 (but we are already at 2.4.7), are there other changes I missed? We will rebuild against 2.4.11.1 (which is the latest in the FreeBSD Ports repos), but it will have little effect if that code has not been touched since 2.4.6.

I just confirmed that we still have the problem with 2.4.11.1 (the latest in the FreeBSD pkg/ports system).

**Comment by pk1048** [comment]
I am also confused as to why the module is still using shared memory when the cache has been configured for memcache.

Lines 597 - 607 of config.c appear to be configuring shared memory *before* the configured cache type has been set in lines 609-618.

**Reply by zandbelt** [reply]
there was another fix for graceful restarts in 2.4.12.1

**Reply by pk1048** [reply]
Thank you, we will give that release a try.
